packages/sanity/src/core/studio/copyPaste/utils.ts

```diff
@@ -3,84 +3,161 @@ import {isString} from 'sanity'
-  try {
-    const clipboardItem: Record<string, Blob> = {
-      [SANITY_CLIPBOARD_ITEM_TYPE]: new Blob([JSON.stringify(copyActionResult)], {
-        type: SANITY_CLIPBOARD_ITEM_TYPE,
+export async function writeClipboardItem(copyActionResult: CopyActionResult): Promise<boolean> {
+  const textValue = transformValueToText(copyActionResult.items[0].value)
+  const escapedTextValue = escapeHtml(textValue)
+  // we use a utf8-safe base64 encoded string to preserve the data as safely as
+  // possible when serializing into HTML
+  const base64SanityClipboardItem = utf8ToBase64(JSON.stringify(copyActionResult))
+
+  const clipboardItem = new ClipboardItem({
+    ...(SUPPORTS_SANITY_CLIPBOARD_MIMETYPE && {
+      [MIMETYPE_SANITY_CLIPBOARD]: new Blob([JSON.stringify(copyActionResult)], {
+        type: MIMETYPE_SANITY_CLIPBOARD,
       }),
-    }
-    if (SANITY_CLIPBOARD_ITEM_TYPE !== 'text/plain') {
-      const text = copyActionResult.items
-        .map((item) => transformValueToText(item.value))
-        .filter(Boolean)
-        .join('\n')
-      if (text.length > 0) {
-        clipboardItem['text/plain'] = new Blob([text], {type: 'text/plain'})
-      }
-    }
-    await window.navigator.clipboard.write([new ClipboardItem(clipboardItem)])
+    }),
+    [MIMETYPE_HTML]: new Blob(
+      // we store the data within a data attribute because safari will sanitize
+      // and mangle the HTML written to the clipboard
+      // https://stackoverflow.com/a/68958287/5776910
+      [`<p data-${BASE64_ATTR}="${base64SanityClipboardItem}">${escapedTextValue}</p>`],
+      {type: MIMETYPE_HTML},
+    ),
+    [MIMETYPE_PLAINTEXT]: new Blob([textValue], {type: MIMETYPE_PLAINTEXT}),
+  })
+
+  try {
+    await navigator.clipboard.write([clipboardItem])
     return true
   } catch (error) {
     console.error(`Failed to write to clipboard: ${error.message}`, error)
     return false
   }
 }
 
-export function transformValueToText(value: unknown): string | number {
-  if (!value) {
-    return ''
+export async function parseClipboardItem(item: ClipboardItem): Promise<CopyActionResult | null> {
+  if (item.types.includes(MIMETYPE_SANITY_CLIPBOARD)) {
+    const blob = await item.getType(MIMETYPE_SANITY_CLIPBOARD)
+    const text = await blob.text()
+    return JSON.parse(text)
   }
-  if (isString(value)) {
-    return value
+
+  if (!item.types.includes(MIMETYPE_HTML)) return null
+  const blob = await item.getType(MIMETYPE_HTML)
+  const html = await blob.text()
+
+  const parser = new DOMParser()
+  const doc = parser.parseFromString(html, 'text/html')
+
+  try {
+    const el = doc.querySelector(`[data-${BASE64_ATTR}]`) as HTMLParagraphElement
+    if (!el) return null
+
+    type CamelCase<S extends string> = S extends `${infer P1}-${infer P2}${infer P3}`
+      ? `${P1}${Capitalize<P2>}${CamelCase<P3>}`
+      : S
+
+    const {sanityClipboardBase64} = el.dataset as Record<CamelCase<typeof BASE64_ATTR>, string>
```

Comment:
Apologies if I'm missing something, but where is "dataset" being set? I'm assuming this is a reference to the <p> tag being injected on line 69 but I'm not quite putting it all together.

Reply:
The `dataset` property is populated from the `DOMParser` from the [`data-` attribute](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes#javascript_access).